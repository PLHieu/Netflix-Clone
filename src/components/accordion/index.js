import React, { createContext, useContext, useState } from 'react';
import { Container, Inner, Frame, Item, Title, Header, Body } from './styles/accordion';

const ToggleFAQContext = createContext({ toggleShow: false, setToggleShow: (toggle) => {} });

export default function Accordion({ children }) {
  return (
    <Container>
      <Inner>{children}</Inner>
    </Container>
  );
}

Accordion.Title = function AccordionTitle({ children, ...restProps }) {
  return <Title {...restProps}>{children}</Title>;
};

Accordion.Frame = function AccordionFrame({ children, ...restProps }) {
  return <Frame {...restProps}>{children}</Frame>;
};

Accordion.Item = function AccordionItem({ children, ...restProps }) {
  const [toggleShow, setToggleShow] = useState(false);
  return (
    <ToggleFAQContext.Provider value={{ toggleShow, setToggleShow }}>
      <Item {...restProps}>{children}</Item>
    </ToggleFAQContext.Provider>
  );
};

Accordion.Header = function AccordionHeader({ children, ...restProps }) {
  const { toggleShow, setToggleShow } = useContext(ToggleFAQContext);
  return (
    <Header onClick={() => setToggleShow(!toggleShow)} {...restProps}>
      {children}
      {toggleShow ? (
        <img src="/images/icons/close-slim.png" alt="Close" />
      ) : (
        <img src="/images/icons/add.png" alt="Open" />
      )}
    </Header>
  );
};

Accordion.Body = function AccordionBody({ children, ...restProps }) {
  const { toggleShow } = useContext(ToggleFAQContext);
  return (
    <Body className={toggleShow ? 'open' : 'closed'} {...restProps}>
      <span>{children}</span>
    </Body>
  );
};
