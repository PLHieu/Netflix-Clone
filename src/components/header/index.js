import { Link as RouterLink } from 'react-router-dom';
import { Background, Logo, Link, Container } from './styles/header';

export default function Header({ children, ...restProps }) {
  return <Background {...restProps}>{children}</Background>;
}

Header.Frame = function HeaderFrame({ children, ...restProps }) {
  return <Container {...restProps}>{children}</Container>;
};

Header.ButtonLink = function HeaderButtonLink({ to, children, ...restProps }) {
  return (
    <Link to={to} {...restProps}>
      {children}
    </Link>
  );
};

Header.Logo = function HeaderLog({ src, to, ...restProps }) {
  return (
    <RouterLink to={to}>
      <Logo src={src} {...restProps} />;
    </RouterLink>
  );
};
