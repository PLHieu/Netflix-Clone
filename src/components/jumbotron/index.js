import React from "react";
import { Container, Image, Inner, Item, Pane, SubTitle, Title } from "./styles/jumbotron";

export default function Jumbotron({
  children,
  direction = "row",
  ...restProps
}) {
  return (
    <div {...restProps}>{children}</div>
  );
}

Jumbotron.Container = function JumbotronContainer({children,...restProps}) {
  return <div {...restProps}>{children}</div>
}

Jumbotron.Pane = function JumbotronPane({children,...restProps}) {
  return <div {...restProps}>{children}</div>
}

Jumbotron.Title = function JumbotronTitle({children,...restProps}) {
  return <h1 {...restProps}>{children}</h1>
}

Jumbotron.Subtitle = function JumbotronSubtitle({children,...restProps}) {
  return <h2 {...restProps}>{children}</h2>
}

Jumbotron.Image = function JumbotronImage({...restProps}) {
  return <img {...restProps} />
}
