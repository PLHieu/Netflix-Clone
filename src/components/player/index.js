import React, { useContext, useState } from 'react';
import ReactDOM from 'react-dom';
import { PlayerContext, SourcePlayerContext } from '../../context/player';
import { Button, Close, Container, Inner, Overlay } from './styles/player';

export default function Player({ children, ...restProps }) {
  const [showPlayer, setShowPlayer] = useState(false);
  return (
    <PlayerContext.Provider value={{ showPlayer, setShowPlayer }}>
      <Container {...restProps}>{children}</Container>
    </PlayerContext.Provider>
  );
}

Player.Video = function PlayerVideo() {
  const { showPlayer, setShowPlayer } = useContext(PlayerContext);
  const { srcPlayer } = useContext(SourcePlayerContext);

  return showPlayer
    ? ReactDOM.createPortal(
        <Overlay onClick={() => setShowPlayer(false)}>
          <Inner>
            <video id="netflix-player" controls>
              <source src={srcPlayer} type="video/mp4" />
            </video>
            <Close />
          </Inner>
        </Overlay>,
        document.body
      )
    : null;
};

Player.Button = function PlayerButton({ ...restProps }) {
  const { showPlayer, setShowPlayer } = useContext(PlayerContext);

  return (
    <Button onClick={() => setShowPlayer(!showPlayer)} {...restProps}>
      Play
    </Button>
  );
};
