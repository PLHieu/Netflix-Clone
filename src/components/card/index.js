import React, { useContext, useState } from 'react';
import { CardFeatureContext } from '../../context/cardFeature';
import { SourcePlayerContext } from '../../context/player';

import {
  Container,
  Group,
  Title,
  SubTitle,
  Text,
  Meta,
  Entities,
  Item,
  Image,
  Feature,
  Content,
  FeatureText,
  FeatureTitle,
  FeatureClose,
  Maturity,
} from './styles/card';

export default function Card({ children, ...restProps }) {
  const [showFeature, setShowFeature] = useState(false);
  const [itemFeature, setItemFeature] = useState({});
  const [srcPlayer, setSrcPlayer] = useState('');
  return (
    <SourcePlayerContext.Provider value={{ srcPlayer, setSrcPlayer }}>
      <CardFeatureContext.Provider value={{ itemFeature, showFeature, setShowFeature, setItemFeature }}>
        <Container {...restProps}>{children}</Container>
      </CardFeatureContext.Provider>
    </SourcePlayerContext.Provider>
  );
}

Card.Group = function CardGroup({ children, ...restProps }) {
  return <Group {...restProps}>{children}</Group>;
};

Card.Title = function CardTitle({ children, ...restProps }) {
  return <Title {...restProps}>{children}</Title>;
};

Card.SubTitle = function CardSubTitle({ children, ...restProps }) {
  return <SubTitle {...restProps}>{children}</SubTitle>;
};

Card.Text = function CardText({ children, ...restProps }) {
  return <Text {...restProps}>{children}</Text>;
};

Card.Entities = function CardEntities({ children, ...restProps }) {
  return <Entities {...restProps}>{children}</Entities>;
};

Card.Meta = function CardMeta({ children, ...restProps }) {
  return <Meta {...restProps}>{children}</Meta>;
};

Card.Item = function CardItem({ item, children, ...restProps }) {
  const { setItemFeature, setShowFeature } = useContext(CardFeatureContext);
  const { setSrcPlayer } = useContext(SourcePlayerContext);
  return (
    <Item
      onClick={() => {
        setItemFeature(item);
        setShowFeature(true);
        setSrcPlayer('/videos/bunny.mp4');
      }}
      {...restProps}
    >
      {children}
    </Item>
  );
};

Card.Image = function CardImage({ ...restProps }) {
  return <Image {...restProps} />;
};

Card.Feature = function CardFeature({ category, children, ...restProps }) {
  const { showFeature, itemFeature, setShowFeature } = useContext(CardFeatureContext);
  return showFeature ? (
    <Feature {...restProps} src={`/images/${category}/${itemFeature.genre}/${itemFeature.slug}/large.jpg`}>
      <Content>
        <FeatureTitle>{itemFeature.title}</FeatureTitle>
        <FeatureText>{itemFeature.description}</FeatureText>
        <FeatureClose onClick={() => setShowFeature(false)}>
          <img src="/images/icons/close.png" alt="close" />
        </FeatureClose>

        <Group margin="30px 0" flexDirection="row" alignItems="center">
          <Maturity rating={itemFeature.maturity}>{itemFeature.maturity < 12 ? 'PG' : itemFeature.maturity}</Maturity>
          <FeatureText fontWeight="bold">
            {itemFeature.genre.charAt(0).toUpperCase() + itemFeature.genre.slice(1)}
          </FeatureText>
        </Group>
        {children}
      </Content>
    </Feature>
  ) : null;
};
