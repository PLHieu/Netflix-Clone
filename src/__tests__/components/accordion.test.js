import { fireEvent, render } from '@testing-library/react';
import React from 'react';
import { Accordion } from '../../components';
import faqData from '../../fixtures/faqs.json';

describe('Accordion component', () => {
  it('render the <Accordion /> with populated data', () => {
    const { getByText } = render(
      <Accordion>
        <Accordion.Title>Frequently Asked Questions</Accordion.Title>
        <Accordion.Frame>
          {faqData.map((item) => (
            <Accordion.Item key={item.id}>
              <Accordion.Header>{item.header}</Accordion.Header>
              <Accordion.Body>{item.body}</Accordion.Body>
            </Accordion.Item>
          ))}
        </Accordion.Frame>
      </Accordion>
    );

    expect(getByText('Frequently Asked Questions')).toBeTruthy();
    expect(getByText('What is Netflix?')).toBeTruthy();
    expect(getByText('How much does Netflix cost?')).toBeTruthy();
    expect(getByText('Where can I watch?')).toBeTruthy();
    expect(getByText('How do I cancel?')).toBeTruthy();
    expect(getByText('What can I watch on Netflix?')).toBeTruthy();
  });

  it('Open and Close <Accordion /> Item ', () => {
    const { container, queryByText } = render(
      <Accordion>
        <Accordion.Title>Frequently Asked Questions</Accordion.Title>
        <Accordion.Frame>
          {faqData.map((item) => (
            <Accordion.Item key={item.id}>
              <Accordion.Header>{item.header}</Accordion.Header>
              <Accordion.Body>{item.body}</Accordion.Body>
            </Accordion.Item>
          ))}
        </Accordion.Frame>
      </Accordion>
    );

    const whatIsNetflixBodyText =
      "Netflix is a streaming service that offers a wide variety of award-winning TV programmes, films, anime, documentaries and more – on thousands of internet-connected devices. You can watch as much as you want, whenever you want, without a single advert – all for one low monthly price. There's always something new to discover, and new TV programmes and films are added every week!";

    expect(queryByText(whatIsNetflixBodyText).closest('div').classList.contains('closed')).toBe(true);
    fireEvent.click(queryByText('What is Netflix?'));
    expect(queryByText(whatIsNetflixBodyText).closest('div').classList.contains('open')).toBe(true);

    fireEvent.click(queryByText('What is Netflix?'));
    expect(queryByText(whatIsNetflixBodyText).closest('div').classList.contains('closed')).toBe(true);

    expect(container.firstChild).toMatchSnapshot();
  });
});
