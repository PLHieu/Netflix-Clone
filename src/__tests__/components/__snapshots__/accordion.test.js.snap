// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion component Open and Close <Accordion /> Item  1`] = `
<div
  class="accordion__Container-sc-ne36y7-0 CXSjt"
>
  <div
    class="accordion__Inner-sc-ne36y7-2 dCjPiB"
  >
    <h1
      class="accordion__Title-sc-ne36y7-3 gaxEyC"
    >
      Frequently Asked Questions
    </h1>
    <div
      class="accordion__Frame-sc-ne36y7-1 jksrMQ"
    >
      <div
        class="accordion__Item-sc-ne36y7-4 hCGVJk"
      >
        <div
          class="accordion__Header-sc-ne36y7-5 gTkrWM"
        >
          What is Netflix?
          <img
            alt="Open"
            src="/images/icons/add.png"
          />
        </div>
        <div
          class="accordion__Body-sc-ne36y7-6 gSJQcQ closed"
        >
          <span>
            Netflix is a streaming service that offers a wide variety of award-winning TV programmes, films, anime, documentaries and more – on thousands of internet-connected devices.

You can watch as much as you want, whenever you want, without a single advert – all for one low monthly price. There's always something new to discover, and new TV programmes and films are added every week!
          </span>
        </div>
      </div>
      <div
        class="accordion__Item-sc-ne36y7-4 hCGVJk"
      >
        <div
          class="accordion__Header-sc-ne36y7-5 gTkrWM"
        >
          How much does Netflix cost?
          <img
            alt="Open"
            src="/images/icons/add.png"
          />
        </div>
        <div
          class="accordion__Body-sc-ne36y7-6 gSJQcQ closed"
        >
          <span>
            Watch Netflix on your smartphone, tablet, smart TV, laptop or streaming device, all for one low fixed monthly fee. Plans start from £5.99 a month. No extra costs or contracts.
          </span>
        </div>
      </div>
      <div
        class="accordion__Item-sc-ne36y7-4 hCGVJk"
      >
        <div
          class="accordion__Header-sc-ne36y7-5 gTkrWM"
        >
          Where can I watch?
          <img
            alt="Open"
            src="/images/icons/add.png"
          />
        </div>
        <div
          class="accordion__Body-sc-ne36y7-6 gSJQcQ closed"
        >
          <span>
            Watch anywhere, anytime, on an unlimited number of devices. Sign in with your Netflix account to watch instantly on the web at netflix.com from your personal computer or on any internet-connected device that offers the Netflix app, including smart TVs, smartphones, tablets, streaming media players and game consoles.

You can also download your favourite programmes with the iOS, Android, or Windows 10 app. Use downloads to watch while you're on the go and without an internet connection. Take Netflix with you anywhere.
          </span>
        </div>
      </div>
      <div
        class="accordion__Item-sc-ne36y7-4 hCGVJk"
      >
        <div
          class="accordion__Header-sc-ne36y7-5 gTkrWM"
        >
          How do I cancel?
          <img
            alt="Open"
            src="/images/icons/add.png"
          />
        </div>
        <div
          class="accordion__Body-sc-ne36y7-6 gSJQcQ closed"
        >
          <span>
            Netflix is flexible. There are no annoying contracts and no commitments. You can easily cancel your account online in two clicks. There are no cancellation fees – start or stop your account at any time.
          </span>
        </div>
      </div>
      <div
        class="accordion__Item-sc-ne36y7-4 hCGVJk"
      >
        <div
          class="accordion__Header-sc-ne36y7-5 gTkrWM"
        >
          What can I watch on Netflix?
          <img
            alt="Open"
            src="/images/icons/add.png"
          />
        </div>
        <div
          class="accordion__Body-sc-ne36y7-6 gSJQcQ closed"
        >
          <span>
            Netflix has an extensive library of feature films, documentaries, TV programmes, anime, award-winning Netflix originals, and more. Watch as much as you want, any time you want.
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
